qualtran/bloqs/chemistry/ising/walk_operator.py

```diff
@@ -0,0 +1,171 @@
+from qualtran.bloqs.state_preparation import StatePreparationAliasSampling
+from qualtran.symbolics import SymbolicFloat, SymbolicInt
+
+
+def get_prepare_precision_from_eigenphase_precision(
+    eps_eigenphase: SymbolicFloat,
+    num_coeffs: SymbolicInt,
+    sum_of_coeffs: SymbolicFloat,
+    hamiltonian_l2_norm: SymbolicFloat,
+) -> SymbolicFloat:
+    r"""Precision of LCU coefficients to get an `eps_eigenphase` approx. of eigenphases.
+
+    For the overall block-encoding to have precision $\epsilon$, it requires approximating
+    $w_l$ to a precision of $\delta$ such that
+
+    $$
+        \delta = \frac{\epsilon}{(1 + \epsilon^2)L}
+        \left( 1 - \frac{\norm{H}^2}{\lambda^2} \right)
+    $$
+
+    as described in Eq. A9 of [1].
+
+    Args:
+        num_coeffs: number of LCU coefficients $L$.
+        eps_eigenphase: precision to approximate the eigenphases $\epsilon$.
+        sum_of_coeffs: sum of LCU coefficients $\lambda$.
+        hamiltonian_l2_norm: upper bound on $\norm{H}$.
+
+    References:
+        [Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity](https://arxiv.org/abs/1805.03662).
+        Babbush et. al. (2018). Eq (A9).
+    """
+    return ((eps_eigenphase * sum_of_coeffs) / ((1 + eps_eigenphase**2) * num_coeffs)) * (
+        1 - (hamiltonian_l2_norm / sum_of_coeffs) ** 2
+    )
+
+
+def upper_bound_norm_for_pauli_hamiltonian(ham: cirq.PauliSum) -> float:
+    r"""Compute a weak upper-bound of the norm of a PauliSum Hamiltonian.
+
+    Given a Pauli Hamiltonian $H = \sum_{l = 0}^{L-1} w_l H_l$, where each $H_l$
+    is a PauliString, compute an upper bound on $\norm{H}$ which is slightly
+    better than $\lambda = \sum_l w_l$, if possible.
+
+    If all the terms commute, this just returns $\lambda$.
+
+    Proof:
+
+    As all the $H_l$ are Pauli strings, we know
+    - $H_l^2 = I$
+    - $H_l H_{l'} = \pm H_{l'} H_l$ (i.e. they either commute/anti-commute)
+
+    $$
+        H^2
+        = \sum_{l, l'} w_l w_{l'} H_l H_{l'}
+        = \sum_{l} w_l^2 I + \sum_{l < l'; [H_l, H_{l'}] = 0} 2 w_l w_{l'} H_l H_{l'}
+    $$
+
+    $$
+        \norm{H^2}
+        \le \sum_{l} w_l^2 + \sum_{l < l'; [H_l, H_{l'}] = 0} 2 w_l w_{l'}
+        = \lambda^2 - \sum_{l < l'; [H_l, H_{l'}] \ne 0} 2 w_l w_{l'}
+    $$
+
+    Notes:
+        Takes time $L^2 n$, for $L$ terms and each PauliString acting on $n$ qubits.
+    """
+    L = len(ham)
+    ham_ps = [ps for ps in ham]
+    qlambda = sum(ps.coefficient for ps in ham)
+
+    anticommute_sum = 0
+    for i in range(L):
+        for j in range(i):
+            if not cirq.commutes(ham_ps[i], ham_ps[j]):
+                anticommute_sum += np.abs(ham_ps[i].coefficient * ham_ps[j].coefficient)
+
+    return np.sqrt(qlambda**2 - 2 * anticommute_sum)
+
+
+def walk_operator_for_pauli_hamiltonian(
+    ham: cirq.PauliSum, eps: float, *, ham_norm_upper_bound: Optional[float] = None
```

Comment:
Should `eps` be a `SymbolicFloat` ?

Reply:
good point. The current impl doesn't support it, but it's a small change, let me fix that.

Reply:
added support for symbolic eps and ham_norm. Making J, Gamma symbolic would require more work, as `cirq.PauliSum` does not support symbolic coeffs. Deferring to a later PR.